File: apps/mobile-native/android/app/src/main/java/mx/nexara/mobile/nativeapp/access/PermissionLabels.kt
package mx.nexara.mobile.nativeapp.access

/**
 * Traduce claves de permiso (`activities.manage`, `workflow.view`…) a lo que la persona entiende:
 * «Actividades: ver, administrar». Agrupa por módulo y separa lo que existe en NEXARA Core de
 * los permisos heredados de paneles que ya no están en la app.
 */
object PermissionLabels {

    data class Modulo(val clave: String, val nombre: String, val core: Boolean)

    data class Grupo(val modulo: Modulo, val acciones: List<String>)

    /** Prefijos más largos primero para que `company.settings` gane a `company`. */
    private val MODULOS: List<Modulo> = listOf(
        Modulo("company.settings", "Datos de la empresa", true),
        Modulo("executive.dashboard", "Tablero de dirección", true),
        Modulo("sales.targets", "Metas de ventas", false),
        Modulo("hr.approve", "Recursos humanos", true),
        Modulo("activities", "Actividades", true),
        Modulo("evidences", "Evidencias", true),
        Modulo("attendance", "Asistencia", true),
        Modulo("lunch_breaks", "Hora de comida", true),
        Modulo("gps", "Ubicación del equipo", true),
        Modulo("users", "Usuarios", true),
        Modulo("roles", "Roles y permisos", true),
        Modulo("console", "Consola de administración", true),
        Modulo("audit", "Bitácora de cambios", true),
        Modulo("hr", "Recursos humanos", true),
        Modulo("people", "Personal", true),
        Modulo("documents", "Documentos", true),
        Modulo("search", "Búsqueda", true),
        Modulo("clients", "Clientes", true),
        Modulo("chat", "Chat", true),
        Modulo("workflow", "Flujos de aprobación", false),
        Modulo("kb", "Base de conocimiento", false),
        Modulo("support", "Mesa de soporte", false),
        Modulo("panel", "Paneles de la plataforma", false),
        Modulo("cvs", "Revisión de CV", false),
        Modulo("sales", "Ventas", false),
        Modulo("cotizaciones", "Cotizaciones", false),
        Modulo("catalog", "Catálogo", false),
        Modulo("contabilidad", "Contabilidad", false),
        Modulo("accounting", "Contabilidad", false),
        Modulo("invoicing", "Facturación", false),
        Modulo("banking", "Bancos", false),
        Modulo("procurement", "Compras", false),
        Modulo("warehouse", "Almacén", false),
        Modulo("stock", "Inventario", false),
        Modulo("vehicles", "Vehículos", false),
        Modulo("tools", "Herramientas", false),
        Modulo("viatics", "Viáticos", false),
        Modulo("maintenance", "Mantenimiento", false),
        Modulo("assets", "Activos", false),
        Modulo("noc", "Centro de monitoreo", false),
        Modulo("lab", "Laboratorio", false),
        Modulo("bi", "Inteligencia de negocio", false),
    ).sortedByDescending { it.clave.length }

    private val ACCIONES: Map<String, String> = mapOf(
        "view" to "ver",
        "manage" to "administrar",
        "create" to "crear",
        "export" to "exportar",
        "review" to "revisar",
        "approve" to "aprobar",
        "request" to "solicitar",
        "access" to "entrar",
        "admin" to "administrar",
        "inventory" to "inventario",
        "leave" to "aprobar permisos y vacaciones",
        "superadmin.review" to "revisión de dirección",
        "admin.review" to "revisión de administración",
        "support" to "soporte",
        "people" to "personal",
        "noc" to "monitoreo",
        "lab" to "laboratorio",
    )

    fun accion(resto: String): String =
        ACCIONES[resto] ?: resto.split('.', '_').joinToString(" ") { it.lowercase() }

    /** Grupos en orden: primero los de Core, cada uno con sus acciones sin repetir. */
    fun agrupar(permisos: List<String>): List<Grupo> {
        val porModulo = linkedMapOf<String, Pair<Modulo, LinkedHashSet<String>>>()
        for (raw in permisos.map { it.trim() }.filter { it.isNotEmpty() }.distinct()) {
            val modulo = MODULOS.firstOrNull { raw == it.clave || raw.startsWith(it.clave + ".") }
                ?: Modulo(raw.substringBefore('.'), raw.substringBefore('.').replaceFirstChar { it.uppercase() }, false)
            val resto = raw.removePrefix(modulo.clave).trimStart('.').ifBlank { "view" }
            val entrada = porModulo.getOrPut(modulo.nombre) { modulo to linkedSetOf() }
            entrada.second += accion(resto)
        }
        return porModulo.values
            .map { (m, acciones) -> Grupo(m, acciones.toList()) }
            .sortedWith(compareByDescending<Grupo> { it.modulo.core }.thenBy { it.modulo.nombre })
    }

    /** «Ver, administrar y exportar». */
    fun unirAcciones(acciones: List<String>): String = when (acciones.size) {
        0 -> ""
        1 -> acciones[0].replaceFirstChar { it.uppercase() }
        else -> (acciones.dropLast(1).joinToString(", ") + " y " + acciones.last()).replaceFirstChar { it.uppercase() }
    }
}

File: apps/mobile-native/ios/NexaraApp/Access/PermissionLabels.swift
import Foundation

/// Traduce claves de permiso (`activities.manage`, `workflow.view`…) a lo que la
/// persona entiende: «Actividades: ver, administrar». Agrupa por módulo y separa lo
/// que existe dentro de NEXARA Core de lo heredado de paneles que ya no están en la
/// app.
///
/// Espejo de `PermissionLabels` de Android: mismos módulos, mismos verbos, mismo
/// orden. Si allá se añade una clave, aquí también; si no, las dos apps le dirían a
/// la misma persona cosas distintas sobre lo que puede hacer.
enum PermissionLabels {

    struct Modulo: Equatable {
        let clave: String
        let nombre: String
        /// ¿Vive dentro de la app (Core) o es de un panel que solo existe en la web?
        let core: Bool
    }

    struct Grupo: Identifiable {
        let modulo: Modulo
        let acciones: [String]
        /// El nombre del módulo ya es único: se agrupa por él, no por la clave
        /// («Recursos humanos» junta `hr` y `hr.approve`).
        var id: String { modulo.nombre }
    }

    /// Prefijos más largos primero, para que `company.settings` gane a `company`.
    private static let modulos: [Modulo] = [
        Modulo(clave: "company.settings", nombre: "Datos de la empresa", core: true),
        Modulo(clave: "executive.dashboard", nombre: "Tablero de dirección", core: true),
        Modulo(clave: "sales.targets", nombre: "Metas de ventas", core: false),
        Modulo(clave: "hr.approve", nombre: "Recursos humanos", core: true),
        Modulo(clave: "activities", nombre: "Actividades", core: true),
        Modulo(clave: "evidences", nombre: "Evidencias", core: true),
        Modulo(clave: "attendance", nombre: "Asistencia", core: true),
        Modulo(clave: "lunch_breaks", nombre: "Hora de comida", core: true),
        Modulo(clave: "gps", nombre: "Ubicación del equipo", core: true),
        Modulo(clave: "users", nombre: "Usuarios", core: true),
        Modulo(clave: "roles", nombre: "Roles y permisos", core: true),
        Modulo(clave: "console", nombre: "Consola de administración", core: true),
        Modulo(clave: "audit", nombre: "Bitácora de cambios", core: true),
        Modulo(clave: "hr", nombre: "Recursos humanos", core: true),
        Modulo(clave: "people", nombre: "Personal", core: true),
        Modulo(clave: "documents", nombre: "Documentos", core: true),
        Modulo(clave: "search", nombre: "Búsqueda", core: true),
        Modulo(clave: "clients", nombre: "Clientes", core: true),
        Modulo(clave: "chat", nombre: "Chat", core: true),
        Modulo(clave: "workflow", nombre: "Flujos de aprobación", core: false),
        Modulo(clave: "kb", nombre: "Base de conocimiento", core: false),
        Modulo(clave: "support", nombre: "Mesa de soporte", core: false),
        Modulo(clave: "panel", nombre: "Paneles de la plataforma", core: false),
        Modulo(clave: "cvs", nombre: "Revisión de CV", core: false),
        Modulo(clave: "sales", nombre: "Ventas", core: false),
        Modulo(clave: "cotizaciones", nombre: "Cotizaciones", core: false),
        Modulo(clave: "catalog", nombre: "Catálogo", core: false),
        Modulo(clave: "contabilidad", nombre: "Contabilidad", core: false),
        Modulo(clave: "accounting", nombre: "Contabilidad", core: false),
        Modulo(clave: "invoicing", nombre: "Facturación", core: false),
        Modulo(clave: "banking", nombre: "Bancos", core: false),
        Modulo(clave: "procurement", nombre: "Compras", core: false),
        Modulo(clave: "warehouse", nombre: "Almacén", core: false),
        Modulo(clave: "stock", nombre: "Inventario", core: false),
        Modulo(clave: "vehicles", nombre: "Vehículos", core: false),
        Modulo(clave: "tools", nombre: "Herramientas", core: false),
        Modulo(clave: "viatics", nombre: "Viáticos", core: false),
        Modulo(clave: "maintenance", nombre: "Mantenimiento", core: false),
        Modulo(clave: "assets", nombre: "Activos", core: false),
        Modulo(clave: "noc", nombre: "Centro de monitoreo", core: false),
        Modulo(clave: "lab", nombre: "Laboratorio", core: false),
        Modulo(clave: "bi", nombre: "Inteligencia de negocio", core: false),
    ].sorted { $0.clave.count > $1.clave.count }

    private static let verbos: [String: String] = [
        "view": "ver",
        "manage": "administrar",
        "create": "crear",
        "export": "exportar",
        "review": "revisar",
        "approve": "aprobar",
        "request": "solicitar",
        "access": "entrar",
        "admin": "administrar",
        "inventory": "inventario",
        "leave": "aprobar permisos y vacaciones",
        "superadmin.review": "revisión de dirección",
        "admin.review": "revisión de administración",
        "support": "soporte",
        "people": "personal",
        "noc": "monitoreo",
        "lab": "laboratorio",
    ]

    /// Lo que queda de la clave después del módulo, en palabras. Sin traducción
    /// conocida se enseña tal cual, con puntos y guiones bajos vueltos espacios:
    /// mejor un verbo raro que esconderle a alguien un permiso que sí tiene.
    static func accion(_ resto: String) -> String {
        if let conocido = verbos[resto] { return conocido }
        return resto
            .split(whereSeparator: { $0 == "." || $0 == "_" })
            .map { $0.lowercased() }
            .joined(separator: " ")
    }

    /// Grupos en orden: primero los de Core, cada uno con sus acciones sin repetir.
    static func agrupar(_ permisos: [String]) -> [Grupo] {
        var ordenDeAparicion: [String] = []
        var porModulo: [String: (modulo: Modulo, acciones: [String])] = [:]
        var yaVistos = Set<String>()

        for crudo in permisos.map({ $0.trimmingCharacters(in: .whitespacesAndNewlines) }) {
            guard !crudo.isEmpty, yaVistos.insert(crudo).inserted else { continue }

            let modulo = modulos.first { crudo == $0.clave || crudo.hasPrefix($0.clave + ".") }
                ?? desconocido(crudo)

            var resto = String(crudo.dropFirst(modulo.clave.count))
            while resto.hasPrefix(".") { resto.removeFirst() }
            // `chat` a secas es «chat: ver», no «chat: chat».
            if resto.isEmpty { resto = "view" }
            let etiqueta = accion(resto)

            if var entrada = porModulo[modulo.nombre] {
                if !entrada.acciones.contains(etiqueta) { entrada.acciones.append(etiqueta) }
                porModulo[modulo.nombre] = entrada
            } else {
                ordenDeAparicion.append(modulo.nombre)
                porModulo[modulo.nombre] = (modulo, [etiqueta])
            }
        }

        return ordenDeAparicion
            .compactMap { porModulo[$0] }
            .map { Grupo(modulo: $0.modulo, acciones: $0.acciones) }
            .sorted { izq, der in
                // Core arriba; dentro de cada bloque, por nombre. El nombre es
                // único, así que el orden no depende de cómo llegó la lista.
                if izq.modulo.core != der.modulo.core { return izq.modulo.core }
                return izq.modulo.nombre.localizedCaseInsensitiveCompare(der.modulo.nombre) == .orderedAscending
            }
    }

    /// Permiso de un módulo que no está en la tabla: se le pone el prefijo como
    /// nombre y se trata como de fuera de la app.
    private static func desconocido(_ crudo: String) -> Modulo {
        let prefijo = String(crudo.prefix(while: { $0 != "." }))
        return Modulo(clave: prefijo, nombre: conMayuscula(prefijo), core: false)
    }

    /// «Ver, administrar y exportar».
    static func unirAcciones(_ acciones: [String]) -> String {
        switch acciones.count {
        case 0:
            return ""
        case 1:
            return conMayuscula(acciones[0])
        default:
            let todasMenosLaUltima = acciones.dropLast().joined(separator: ", ")
            return conMayuscula("\(todasMenosLaUltima) y \(acciones[acciones.count - 1])")
        }
    }

    private static func conMayuscula(_ texto: String) -> String {
        guard let primera = texto.first else { return texto }
        return primera.uppercased() + texto.dropFirst()
    }
}
